Feat(datapipes): Zarr mesh schema and readers — a zarr storage path for mesh-native CAE pipelines (#1798)
NVIDIA/physicsnemo: physicsnemo/datapipes/MESH_ZARR_SCHEMA.md

<!-- markdownlint-disable MD013 -->
# PhysicsNeMo Mesh-Zarr Schema

Version: **1** (`schema_version` attr). Reference implementation:
`physicsnemo.datapipes.save_mesh_to_zarr` / `save_domain_mesh_to_zarr`;
validation: `physicsnemo.datapipes.validate_mesh_zarr`. Emitted by
PhysicsNeMo-Curator's `DomainMeshZarrSink`; consumed by `ZarrMeshReader` /
`ZarrDomainMeshReader`.

> ## Mesh group (`format = "physicsnemo-mesh-zarr"`)

Reviewer: ```suggestion
## Mesh Group (`format = "physicsnemo-mesh-zarr"`)
```

One zarr group per mesh:

```text
<group>/                      attrs (required): format, schema_version,
    points                    #   n_points, n_cells; (when n_cells > 0):
    cells                     #   nodes_per_cell, layout
    point_data/<field>        # optional subgroup, one array per field
    cell_data/<field>         # optional subgroup
    global_data/<field>       # optional subgroup (0-d and small arrays)
```

| Array | Shape | Notes |
|---|---|---|
| `points` | `(n_points, n_spatial_dims)` | required |
| `cells` | `(n_cells, nodes_per_cell)` | absent for point clouds (`n_cells == 0`); integer dtype; indices into `points` |
| `point_data/<f>` | `(n_points, ...)` | array or subgroup (see below) |
| `cell_data/<f>` | `(n_cells, ...)` | array or subgroup (see below) |
| `global_data/<f>` | any (typically 0-d) | array or subgroup (see below) |

> ### Field names and nesting

Reviewer: ```suggestion
### Field Names and Nesting
```

> Data fields mirror the `TensorDict` tree: each `<f>` under `point_data` /

Reviewer: ```suggestion
Data fields mirror the `TensorDict` tree. Each `<f>` under `point_data`,
```

> `cell_data` / `global_data` is either an **array** (a tensor leaf) or a

Reviewer: ```suggestion
`cell_data`, or `global_data` is an *array* (a tensor leaf) or a
```

> **subgroup** (a nested `TensorDict`), recursively. Every *leaf array*

Reviewer: ```suggestion
*subgroup* (a nested `TensorDict`), recursively. Every *leaf array*
```

under `point_data` (resp. `cell_data`) must obey the leading-dimension
rule of its container, at any depth.

> - Field and subgroup names must be non-empty and **must not contain `/`**

Reviewer: ```suggestion
- Field and subgroup names must be non-empty and *must not contain `/`*
```

>   (the zarr path separator -- it would create hierarchy instead of a

Reviewer: ```suggestion
  (the Zarr path separator would create hierarchy instead of a
```

>   field). Writers MUST reject such names; readers may treat any name

Reviewer: ```suggestion
  field). Writers MUST reject such names. Readers can treat any name
```

  reaching them as literal.

> - Non-tensor leaves (`NonTensorData`, e.g. strings) are **not

Reviewer: ```suggestion
- Non-tensor leaves (`NonTensorData`, for example, strings) are *not
```

>   representable in v1**; writers MUST reject them (an attrs-based

Reviewer: ```suggestion
  representable in v1*. Writers MUST reject them because an attrs-based
```

>   encoding is reserved for v2).

Reviewer: ```suggestion
  encoding is reserved for v2.
```

> - Groups MUST be **zarr format 3** (`zarr.json` metadata). Readers MAY

Reviewer: ```suggestion
- Groups MUST be *Zarr format 3* (`zarr.json` metadata). Readers CAN
```

  reject format-2 stores.

### Attributes

- `format` (str, required): `"physicsnemo-mesh-zarr"`.
- `schema_version` (int, required): this document's version.
- `n_points`, `n_cells` (int, required): must equal the array shapes.
- `nodes_per_cell` (int, required when `n_cells > 0`).
- `layout` (str, required when `n_cells > 0`): `"soup"` or `"indexed"`.
  **Writers MUST verify** `layout="soup"` (cells exactly
  `arange(n_points).reshape(n_cells, nodes_per_cell)`) before writing it;
  readers rely on it to skip the `cells` read and synthesize indices.
  `"soup"` groups sacrifice shared-vertex connectivity for contiguous
  block reads -- valid only for cell-centric consumers. Consumers that

>   need connectivity (e.g. graph networks) MUST require `"indexed"`.

Reviewer: ```suggestion
  need connectivity (for example, graph networks) MUST require `"indexed"`.
```

> ## DomainMesh group (`format = "physicsnemo-domainmesh-zarr"`)

Reviewer: ```suggestion
## DomainMesh Group (`format = "physicsnemo-domainmesh-zarr"`)
```

> One zarr group per case, mirroring the `.pdmsh` tree:

Reviewer: ```suggestion
One Zarr group per case, mirroring the `.pdmsh` tree:
```

```text
<case>.zarr/                  attrs (required): format, schema_version,
    global_data/<field>       #   boundary_names (list[str])
    interior/                 # a mesh-schema group
    boundaries/<name>/        # mesh-schema groups, one per boundary_names entry
```

> Case-level `global_data` is **single-sourced at the root** and is not

Reviewer: ```suggestion
Case-level `global_data` is *single-sourced at the root* and is not
```

> copied into member meshes; readers merge it at read time

Reviewer: ```suggestion
copied into member meshes. Readers merge it at read time
```

(`merge_global_data_from`), and a key present both case-level and on a
member mesh is a data error (readers raise). Member meshes keep their own

> mesh-local `global_data` (e.g. `TimeValue`).

Reviewer: ```suggestion
mesh-local `global_data` (for example, `TimeValue`).
```

> ## Versioning policy

Reviewer: ```suggestion
## Versioning Policy
```

> 

Reviewer: ```suggestion

The file format follows these rules for `schema_version`:

```

> - `schema_version` is a single integer; any change that could make a

Reviewer: ```suggestion
- `schema_version` is a single integer. Any change that could make a
```

  correct v(N) reader misread the data increments it.

> - Readers MUST reject groups with `schema_version` greater than the

Reviewer: ```suggestion
- Readers MUST reject groups with a `schema_version` greater than the
```

>   version they implement (raise, never best-effort). Groups without the

Reviewer: ```suggestion
  version they implement (raise an error, never best-effort). Groups without the
```

>   attr are treated as version 1.

Reviewer: ```suggestion
  attribute are treated as version 1.
```

- Writers always stamp the current version.

> ## Reserved for version 2

Reviewer: ```suggestion
## Reserved for Version 2
```

> 
> - **Ragged / mixed-element connectivity**: v1 requires fixed-width

Reviewer: ```suggestion

Version 1 and planned version 2 differ in these areas:

- **Ragged and mixed-element connectivity**: v1 requires fixed-width
```

Reviewer: i don't know what i did to make this act weird

  `cells (n_cells, nodes_per_cell)` (homogeneous simplices). v2 will add

>   an offsets-based encoding (`connectivity` + `offsets` [+ `types`]),

Reviewer: ```suggestion
  an offsets-based encoding (`connectivity` , `offsets`, and optionally `types`),
```

  following VTKHDF/UGRID practice, for tetrahedra/hexahedra/prism volume

>   meshes; the fixed-width form remains the fast path.

Reviewer: ```suggestion
  meshes. The fixed-width form remains the fast path.
```

> - **Connectivity dtype**: v1 preserves the writer's dtype; a v2 writer

Reviewer: ```suggestion
- **Connectivity dtype**: v1 preserves the writer's dtype. A v2 writer
```

>   MAY downcast to the smallest safe integer type.

Reviewer: ```suggestion
  CAN downcast to the smallest safe integer type.
```

- **Non-tensor metadata**: an attrs-based (JSON) encoding for

>   `NonTensorData` leaves (case names, solver strings); v1 writers reject

Reviewer: ```suggestion
  `NonTensorData` leaves (case names, solver strings). v1 writers reject
```

  them.

> ## Design notes (non-normative)

Reviewer: ```suggestion
## Design Notes (non-normative)
```

> 

Reviewer: ```suggestion

Storage layout follows these conventions:

```

> - Chunks are sized along the cell/point axes; align `chunk_cells` with

Reviewer: ```suggestion
- Chunks are sized along the cell or point axes. Align `chunk_cells` with
```

  the training reader's `subsample_n_cells` so one sample read touches at
  most two chunks per field.

> - Default compression is blosc-zstd (bitshuffle); uncompressed stores

Reviewer: ```suggestion
- Default compression is blosc-zstd (bitshuffle). Uncompressed stores
```

  trade bytes for zero decode cost and only win when the working set fits
  the page cache.

> - Groups are self-contained (e.g. SDF source geometry stored as an

Reviewer: ```suggestion
- Groups are self-contained (for example, SDF source geometry stored as an
```

<!-- markdownlint-disable MD013 -->
# PhysicsNeMo Mesh-Zarr Schema

Version: **1** (`schema_version` attr). Reference implementation:
`physicsnemo.datapipes.save_mesh_to_zarr` / `save_domain_mesh_to_zarr`;
validation: `physicsnemo.datapipes.validate_mesh_zarr`. Emitted by
PhysicsNeMo-Curator's `DomainMeshZarrSink`; consumed by `ZarrMeshReader` /
`ZarrDomainMeshReader`.

## Mesh group (`format = "physicsnemo-mesh-zarr"`)

One zarr group per mesh:

```
<group>/                      attrs (required): format, schema_version,
    points                    #   n_points, n_cells; (when n_cells > 0):
    cells                     #   nodes_per_cell, layout
    point_data/<field>        # optional subgroup, one array per field
    cell_data/<field>         # optional subgroup
    global_data/<field>       # optional subgroup (0-d and small arrays)
```

| Array | Shape | Notes |
|---|---|---|
| `points` | `(n_points, n_spatial_dims)` | required |
| `cells` | `(n_cells, nodes_per_cell)` | absent for point clouds (`n_cells == 0`); integer dtype; indices into `points` |
| `point_data/<f>` | `(n_points, ...)` | |
| `cell_data/<f>` | `(n_cells, ...)` | |

> | `global_data/<f>` | any (typically 0-d) | |

Reviewer 2: **[blocker, spec] Legal field names and nesting semantics are undefined.**

The spec never says (a) whether `<field>` may itself be a group (nested TensorDict) or must be an array, (b) what characters a field name may contain — `/` is the zarr path separator and today produces silent data loss (see comment in `zarr_mesh.py`), or (c) how non-tensor metadata (TensorDict `NonTensorData`, e.g. strings) is represented. Since PhysicsNeMo-Curator's `DomainMeshZarrSink` implements this spec independently, these gaps are cross-repo divergence waiting to happen.

Also worth pinning here: the **zarr format version** (v2 vs v3). The readers' group discovery accepts both `zarr.json` and `.zgroup`, but the tensorstore backend hardcodes the `zarr3` driver, so a v2-writing curator would produce stores that validate but fail to read on the preferred backend.

Recommendation: define `<field>` as *array or group, recursively* (mirroring TensorDict), prohibit `/` in names, specify NonTensorData handling (reject in v1 or attrs-encode), and require zarr format 3.

Reviewer 2: Fixed in https://github.com/NVIDIA/physicsnemo/pull/1798/commits/ec56ca9530a775353bdf79052e112d8434bd3a94: `MESH_ZARR_SCHEMA.md` now has a "Field names and nesting" section (array-or-subgroup recursively, `/` prohibited, NonTensorData rejected in v1 with an attrs encoding reserved for v2) and pins zarr format 3.
